pkg/schemacompat/schemacompat.go

```diff
@@ -103,34 +103,36 @@ func stringPointersEqual(p1, p2 *string) bool {
-	errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.Enum, new.Enum, "enum", typeName)...)
+func checkUnsupportedValidationForNumerics(fldPath *field.Path, existing, new *schema.ValueValidation, typeName string) error {
+	err := multierr.Combine(
+		checkUnsupportedValidation(fldPath, existing.Not, new.Not, "not", typeName),
+		checkUnsupportedValidation(fldPath, existing.AllOf, new.AllOf, "allOf", typeName),
+		checkUnsupportedValidation(fldPath, existing.AnyOf, new.AnyOf, "anyOf", typeName),
+		checkUnsupportedValidation(fldPath, existing.OneOf, new.OneOf, "oneOf", typeName),
+		checkUnsupportedValidation(fldPath, existing.Enum, new.Enum, "enum", typeName))
 	if !floatPointersEqual(new.Maximum, existing.Maximum) ||
 		!floatPointersEqual(new.Minimum, existing.Minimum) ||
 		new.ExclusiveMaximum != existing.ExclusiveMaximum ||
 		new.ExclusiveMinimum != existing.ExclusiveMinimum {
-		errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.Maximum, new.Maximum, "maximum", typeName)...)
-		errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.Minimum, new.Minimum, "minimum", typeName)...)
+		err = multierr.Combine(err,
+			checkUnsupportedValidation(fldPath, existing.Maximum, new.Maximum, "maximum", typeName),
+			checkUnsupportedValidation(fldPath, existing.Minimum, new.Minimum, "minimum", typeName))
 	}
 	if !floatPointersEqual(new.MultipleOf, existing.MultipleOf) {
-		errorList = append(errorList, checkUnsupportedValidation(fldPath, existing.MultipleOf, new.MultipleOf, "multipleOf", typeName)...)
+		multierr.AppendInto(&err, checkUnsupportedValidation(fldPath, existing.MultipleOf, new.MultipleOf, "multipleOf", typeName))
 	}
-	return
+	return err
 }
 
-func lcdForStructural(fldPath *field.Path, existing, new *schema.Structural, lcd *schema.Structural, narrowExisting bool) (errorList field.ErrorList) {
+func lcdForStructural(fldPath *field.Path, existing, new *schema.Structural, lcd *schema.Structural, narrowExisting bool) error {
 	if lcd == nil && narrowExisting {
-		return field.ErrorList{field.InternalError(fldPath, errors.New("lcd argument should be passed when narrowExisting is true"))}
+		return field.InternalError(fldPath, errors.New("lcd argument should be passed when narrowExisting is true"))
 	}
 	if new == nil {
-		return field.ErrorList{field.Invalid(fldPath, nil, "new schema doesn't allow anything")}
+		return field.Invalid(fldPath, nil, "new schema doesn't allow anything")
 	}
 	if existing.XPreserveUnknownFields != new.XPreserveUnknownFields {
-		errorList = append(errorList, field.Invalid(fldPath.Child("x-preserve-unknown-fields"), new.XPreserveUnknownFields, "x-preserve-unknown-fields value has been changed in an incompatible way"))
+		return field.Invalid(fldPath.Child("x-preserve-unknown-fields"), new.XPreserveUnknownFields, "x-preserve-unknown-fields value has been changed in an incompatible way")
```

Comment:
This raises a question about the intent of the original code. Before this change, if `x-preserve-unknown-fields` didn't match, we'd append to `errorList`, but because of how named return values work, this would never actually get returned, because all the `return`s below would shadow and override it.

With this change, as it currently stands, if `x-preserve-unknown-fields` doesn't match, we'll return that error only, and not perform LCD for the remaining types.

I think the intention was that if `x-preserve-unknown-fields` didn't match, we should report that, _and_ any other errors from LCDing.

In any case, we should have a test that covers this behavior, whichever behavior we decide we want.